### PR TITLE
Cs/sc 2488 allocate attendee to ate

### PR DESCRIPTION
**Pull request by Charl1996**
[Spec doc](https://docs.google.com/document/d/1hU2SQvnfkfxu3RoxhRCKL6IuWl6zO7nuZFEbgF5grrE/edit#)

Note: This PR compares to [this PR](https://github.com/dimagi/commcare-hq/pull/32630) and is thus marked as a draft as I intend to wait till that PR goes to master to make the official PR -> master.

## Product Description
This PR adds a multiselect widget to the event create form which contains all the "possible attendees" (listed by username) on a domain in the left pane. The user will be able to drag those over to the right pane in order to assign those "possible attendees" to an event. See screenshot below.

image

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-2488)

A new django model is introduced, namely `Attendee`, which stores a reference to the respective `commcare-attendee` cases.
When a user lands on the create event page, the widget containing the list of "possible attendees" will simply list all the cases on the domain with case type `commcare-attendee`. When a "possible attendee" is assigned to an event, an extension case will be created for that case, now with type `commcare-potential-attendee`. Doing it this way allows for a "possible attendee" (i.e. case with type `commcare-attendee`) on a domain to be part of many events through its extension cases. The link to the particular event is stored as a case property on the extension case called `event_id`.

When an attendee is unassigned from an event, the respective extension case will be deleted as well as the case index linking it to the domain `commcare-attendee` case (which represents the actual person).

## Feature Flag
Attendance Tracking

### Safety story
- Did local testing
- Added some tests

### Automated test coverage
Automated tests cover logic for assigning attendees to events

### QA Plan
Link to QA ticket to follow

### Migrations
- [ ] The migrations in this code can be safely applied first independently of the code
The migrations depend on the code.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by SmittieC** [comment]
You got any em' screenshots?

File: corehq/apps/events/models.py [hunk 1]
```diff
@@ -105,16 +115,76 @@ def total_attendance(self):
...
+            only_ids=True
+        )
+
+        attendees_to_assign, attendees_to_unassign = find_difference(
+            current_attendees_ids,
+            attendees_case_ids
+        )
+
+        self._assign_attendees(list(attendees_to_assign))
+        self._unassign_attendees(list(attendees_to_unassign))
```
Comment (sravfeyn) [inline corehq/apps/events/models.py:144]: Are these (do they need to be) idempotent?

File: corehq/apps/events/models.py [hunk 2]
```diff
@@ -129,16 +199,65 @@ def _persist_event(self):
...
+        event_obj.domain = data['domain']
+        event_obj.name = data['name']
+        event_obj.start_date = data['start_date']
+        event_obj.end_date = data['end_date']
+        event_obj.attendance_target = data['attendance_target']
+        event_obj.sameday_reg = data['sameday_reg']
+        event_obj.track_each_day = data['track_each_day']
+        event_obj.manager = data['manager']
+        event_obj.is_open = data.get('is_open', Event.is_open)
+        event_obj.attendee_list_status = data.get('attendee_list_status', Event.attendee_list_status)
```
Comment (sravfeyn) [inline corehq/apps/events/models.py:218]: Is it possible to use `cls(**data)` and modify the ones that are special to be more DRY?

**Comment by sravfeyn** [comment]
Added few comments, but looks great!

File: corehq/apps/events/models.py [hunk 3]
```diff
@@ -50,36 +58,18 @@ class Event:
...
-        event_obj.sameday_reg = data['sameday_reg']
-        event_obj.track_each_day = data['track_each_day']
-        event_obj.manager = data['manager']
-        event_obj.is_open = data.get('is_open', Event.is_open)
-        event_obj.attendee_list_status = data.get('attendee_list_status', Event.attendee_list_status)
-
-        if 'case' in data:
-            event_obj.case = data['case']
+    expected_attendees: set[Attendee] = []
+    attendance_takers: set[CommCareUser] = []
```
Comment (kaapstorm) [inline corehq/apps/events/models.py:64]: Don't set default values to mutable data types. [Bad things](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) can happen.
```suggestion
    expected_attendees: set[Attendee] | None = None
    attendance_takers: set[CommCareUser] | None = None
```

File: corehq/apps/events/models.py [hunk 4]
```diff
@@ -47,17 +67,143 @@ class Meta:
...
+        This method deletes the indices and cases linking the domain attendees to the event
+        """
+        if not attendees_case_ids:
+            return
+
+        extension_cases_ids = CommCareCaseIndex.objects.get_extension_case_ids(
+            self.domain,
+            attendees_case_ids
+        )
+        if not extension_cases_ids:
```
Comment (Charl1996) [inline corehq/apps/events/models.py:137]: When this method has executed once and is called again, `extension_cases_ids` will be empty (as expected). This check simply returns since nothing is left to do.

File: corehq/apps/events/models.py [hunk 5]
```diff
@@ -47,17 +67,143 @@ class Meta:
...
+        )
+
+        self._assign_attendees(list(attendees_to_assign))
+        self._unassign_attendees(list(attendees_to_unassign))
+
+    def _assign_attendees(self, attendees_case_ids):
+        if not attendees_case_ids:
+            return
+
+        # How to make this idempotent?
```
Comment (Charl1996) [inline corehq/apps/events/models.py:107]: Still to think of a way to make this idempotent in a not-too-expensive way.

File: corehq/apps/events/models.py [hunk 6]
```diff
@@ -164,3 +169,33 @@ def get_by_event_id(cls, event_id, domain, only_ids=False) -> set[Attendee]:
             return referenced_case_ids
 
         return cls.get_by_ids(referenced_case_ids, domain)
+
+    def save(self, user_id):
+        if not user_id:
+            raise InvalidAttendee('Attendee must have user_id specified')
```
Comment (kaapstorm) [inline corehq/apps/events/models.py:183]: Attendees don't have to be mobile workers. See "[Assumptions](https://docs.google.com/document/d/1hU2SQvnfkfxu3RoxhRCKL6IuWl6zO7nuZFEbgF5grrE/edit#heading=h.mtfd39lckcwc)" and the user story "[An agricultural training event](https://docs.google.com/document/d/1hU2SQvnfkfxu3RoxhRCKL6IuWl6zO7nuZFEbgF5grrE/edit#heading=h.as5own42v8pq)".

File: corehq/apps/events/models.py [hunk 7]
```diff
@@ -137,28 +144,15 @@ def _unassign_attendees(self, attendees_case_ids):
...
-    ATTENDEE_CASE_TYPE = 'commcare-attendee'
-    EVENT_ATTENDEE_CASE_TYPE = 'commcare-potential-attendee'
-
-    domain = models.CharField(max_length=255)
-    case_id = models.CharField(max_length=126)
-
-    class Meta:
-        db_table = "commcare_attendee"
-        indexes = (models.Index(fields=("domain",)),)
+class AttendeeObjectManager(models.Manager):
```
Comment (kaapstorm) [inline corehq/apps/events/models.py:147]: I like what you've done here. I'm still uncertain about using a Django model to manage `CommCareCase`s though. I'll play around with this a bit.

File: corehq/apps/events/utils.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import uuid
+from corehq.apps.hqcase.utils import submit_case_blocks
+from casexml.apps.case.mock import CaseBlock
+from corehq.form_processor.models import CommCareCaseIndex
+
+
+def create_case_with_case_type(case_type, case_args, index=None):
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/events/utils.py:10]: Fair enough. :+1:

File: corehq/form_processor/models/cases.py [hunk 9]
```diff
@@ -1047,6 +1047,14 @@ def get_extension_chain(self, domain, case_ids, include_closed=True, exclude_for
             all_extension_ids = all_extension_ids | new_extensions
         return all_extension_ids
 
+    def get_by_identifier(self, domain, identifier):
+        for db_name in get_db_aliases_for_partitioned_query():
+            query = self.using(db_name).filter(
+                domain=domain,
+                identifier=identifier,
+            )
+        return list(query)
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/form_processor/models/cases.py:1056]: Oh my goodness! How'd I miss that! (btw, didn't know of @only_run_with_partitioned_database. Seems nifty!) 

Will update!

File: corehq/apps/events/tests/test_models.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    @classmethod
+    def tearDownClass(cls):
+        cls.domain_obj.delete()
+        for attendee in Attendee.objects.by_domain(cls.domain):
+            attendee.delete()
+        super().tearDownClass()
+
+    def test_save_fails_without_user_id(self):
+        try:
+            Attendee(domain='bogus').save(user_id=None)
```
Comment (SmittieC) [inline corehq/apps/events/tests/test_models.py:33]: Nit: We should rather use the assertRaises context manager here:
```python
with self.assertRaises(InvalidAttendee):
     Attendee(domain='bogus').save(user_id=None)

File: corehq/apps/events/models.py [hunk 11]
```diff
@@ -19,8 +29,18 @@
 ]
 
 
+class EventObjectManager(models.Manager):
```
Comment (SmittieC) [inline corehq/apps/events/models.py:32]: Why is this moved from the Event's model itself?

**Comment by Charl1996** [comment]
Closed in favour of [this PR](https://github.com/dimagi/commcare-hq/pull/32690)
